src/neural/metal/network_metal.h

```diff
@@ -0,0 +1,161 @@
+    const auto iter_val =
+        &inputs_outputs_->input_val_mem_[batch_size_ * kInputPlanes];
+
+    int i = 0;
+    for (const auto& plane : input) {
+      iter_mask[i] = plane.mask;
+      iter_val[i] = plane.value;
+      i++;
+    }
+
+    batch_size_++;
+  }
+
+  void ComputeBlocking() override;
+
+  int GetBatchSize() const override { return batch_size_; }
+
+  float GetQVal(int sample) const override {
+    if (wdl_) {
+      auto w = inputs_outputs_->op_value_mem_[3 * sample + 0];
+      auto l = inputs_outputs_->op_value_mem_[3 * sample + 2];
+      return w - l;
+    } else {
+      return inputs_outputs_->op_value_mem_[sample];
+    }
+  }
+
+  float GetDVal(int sample) const override {
+    if (wdl_) {
+      auto d = inputs_outputs_->op_value_mem_[3 * sample + 1];
+      return d;
+    } else {
+      return 0.0f;
+    }
+  }
+
+  float GetPVal(int sample, int move_id) const override {
+    return inputs_outputs_->op_policy_mem_[sample * kNumOutputPolicy + move_id];
+  }
+
+  float GetMVal(int sample) const override {
+    if (moves_left_) {
+      return inputs_outputs_->op_moves_left_mem_[sample];
+    }
+    return 0.0f;
+  }
+
+ private:
+  // Memory holding inputs, outputs.
+  std::unique_ptr<InputsOutputs> inputs_outputs_;
+  int batch_size_;
+  bool wdl_;
+  bool moves_left_;
+
+  MetalNetwork* network_;
+};
+
+class MetalNetwork : public Network {
+ public:
+  MetalNetwork(const WeightsFile& file, const OptionsDict& options);
+  ~MetalNetwork() {
+    // if (builder_) { /** @todo clean-up delegate first */ delete builder; builder = NULL; }
+  }
+
+  void forwardEval(InputsOutputs* io, int inputBatchSize);
+
+  std::unique_ptr<NetworkComputation> NewComputation() override {
+    return std::make_unique<MetalNetworkComputation>(this, wdl_, moves_left_);
+  }
+
+  std::unique_ptr<InputsOutputs> GetInputsOutputs() {
+    std::lock_guard<std::mutex> lock(inputs_outputs_lock_);
+    if (free_inputs_outputs_.empty()) {
+      return std::make_unique<InputsOutputs>(max_batch_size_, wdl_,
+                                             moves_left_, conv_policy_);
+    } else {
+      std::unique_ptr<InputsOutputs> resource =
+          std::move(free_inputs_outputs_.front());
+      free_inputs_outputs_.pop_front();
+      return resource;
+    }
+  }
+
+  void ReleaseInputsOutputs(std::unique_ptr<InputsOutputs> resource) {
+    std::lock_guard<std::mutex> lock(inputs_outputs_lock_);
+    free_inputs_outputs_.push_back(std::move(resource));
+  }
+
+  const NetworkCapabilities& GetCapabilities() const override {
+    return capabilities_;
+  }
+
+ private:
+  NetworkCapabilities capabilities_{
+      pblczero::NetworkFormat::INPUT_CLASSICAL_112_PLANE,
+      pblczero::NetworkFormat::MOVES_LEFT_NONE
+  };
+  int max_batch_size_;
+  int batch_size_;
+  int steps_;
+  bool wdl_;
+  bool moves_left_;
+  bool conv_policy_;
+  std::mutex inputs_outputs_lock_;
+  std::list<std::unique_ptr<InputsOutputs>> free_inputs_outputs_;
```

Comment:
Did you check that free list actually helps here? It feels that the performance benefit won't be noticeable. I suspect that due to mutex+std::list iterator follow, it may be slower than just allocating it every time (especially with TCmalloc).

But if free list is really needed, then std::vector (working as stack rather than queue) is usually better.

Reply:
IIRC this is originally from the cuda backend where also gpu buffers are allocated - usually gpu buffer allocation is more time consuming so it makes sense to reuse them. I used the same approach in the onednn backend: the code in master is mainly for cpu, but #1756 adds igpu buffer reuse for (measured) speed improvements.

Reply:
Yes. This was taken from onednn.